### PR TITLE
[WIP] Bring typings from DefinitelyTyped into this repo

### PR DESCRIPTION
**Pull request by samtsai**
**What**:
Improve support of typings by bringing them into this repo instead of relying on a versioned external repo

**Why**:
Help with stability for consumers using typings and give a clearer understanding of what should be supported in a given version without having to reference multiple repos.

**How**:
Modeled after these changes in a sibling repo: https://github.com/testing-library/dom-testing-library/pull/530

**Checklist**:

- [ ] Documentation added to the
      [docs site](https://github.com/alexkrolick/testing-library-docs)
- [x] Tests
- [x] Typescript definitions updated
- [ ] Ready to be merged
      

Currently this does not validate with several errors for Typescript@4.0, example:

```
ERROR: 6:23   expect      TypeScript@4.0 compile error:
Cannot use JSX unless the '--jsx' flag is provided.
```

I'm still investigating this, but if anyone knows more about `dtslint` I'm all 👂

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// TypeScript Version: 3.8
```
Comment (samtsai) [inline types/index.d.ts:1]: Bumped this from 3.1 to 3.8 for this support:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-exports

**Comment by kentcdodds** [comment]
Looks awesome. I don't know about the dtslint stuff, but once that's worked out I'm good to merge 👍 Thanks!

**Comment by kentcdodds** [comment]
@all-contributors please add @samtsai for code and tests

**Comment by kentcdodds** [comment]
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

**Pull request merged**